src/merkle/tiered_smt/bits.rs

```diff
@@ -0,0 +1,66 @@
+use super::{Felt, RpoDigest, TieredSmt, TreeIndex};
+
+pub struct BitsIterator<'a> {
```

Comment:
I wonder if having this `BitsIterator` is a bit of an overkill for our use case - at least in its current form. For example, we will never need to iterate over more than a single element. So, no need to track the current element index.

It might make sense to roll this into node index struct.

Reply:
The idea was to keep the responsibility of the components separated and minimal, having the node index only providing a map for the tree, and the bits iterator with a state to keep track of the current bit.

Another point to consider is that bits iterator is an internal helper of the implementation that won't be exposed by the API, and the node index will be public. If we move the functionality from the bits iterator to the node index, we might end up increasing the complexity of the public API.

wdyt?

Reply:
I agree with @bobbinth here, the code would be simpler without the iterator. One suggestion to keep the separation is to use something like `trait SMTIndex`

Reply:
We need to keep track of the depth as well. I'll describe in one example. Assume we have a tier depth of `3`, and a node index of `{ depth: 6, index: 36 }`.

The key for such case would be `100100`. It means that we should, for the first iteration, traverse `right, left, left`, and for the next one, again, `right, left, left`.

If we `From<Word>`, we will take the first `depth` bits (that is `3`), and for this case we will not be able to differentiate if we should construct a `NodeIndex { depth: 3: index: 4 }` or `NodeIndex { depth: 6, index: 36 }` because we lose the current depth iteration.

What we can do as alternative to the bits iterator is to remove it entirely and create a method `traverse` on `NodeIndex` that will take the first `3` bits of the provided word, return the shifted word, and mutate the current node index. Something like

`fn traverse(&mut self, key: Word) -> Word`

But then we will expose the need for the user to manage this `Word`, as opposed to having everything encapsulated in the bits iterator. It is a working alternative, if we want to get rid of the bits iterator.